### PR TITLE
Add `compute_density_function` to replace removed `sample_noise

### PR DESCRIPTION
**Pull request by Crec0**
Sorry about accidentally closing #1590

### CODE REVIEW

**Comment by gnembon** [comment]
Thanks for contribution. First question before I get started - why name change? looks like the call does what the previous one was doing?

**Comment by Crec0** [comment]
Hey,
It's not really same as before. Previously it was sampling actual noise values, howerver, now it's computing density functions which are used to sample noises either directly or after applying some operation on them. Datapacks and game internally also refers to them distinctively. Since there's a clear distinction between the two terms, in order to eleminate confusion between the two, the name is was changed to `compute_density_function`

File: src/main/java/carpet/script/api/WorldAccess.java
Comment (altrisi) [inline src/main/java/carpet/script/api/WorldAccess.java:1655]: I believe gnembon doesn't like `var` usage.

File: src/main/java/carpet/script/api/WorldAccess.java
Comment (altrisi) [inline src/main/java/carpet/script/api/WorldAccess.java:1685]: Given the `densityFunctionQuery` is used for mapping the results, why change it here? That means it's possible these won't be the same and it may be confusing. Also I believe ResLoc.toString returns with the `minecraft` namespace, while scarpet usually skips it (see `ValueConversions`)

**Review by altrisi — COMMENTED** [review]
Genuinely forgot about this PR, sorry!

Some minor comments, looks good otherwise.

File: src/main/java/carpet/script/api/WorldAccess.java
Comment (altrisi) [inline src/main/java/carpet/script/api/WorldAccess.java:1645]: `Stream.toList()` creates an immutable list, while Scarpet lists should be mutable (and `wrap` doesn't copy the list).
```suggestion
            return ListValue.wrap(densityFunctionRegistry.keySet().stream().map(ResourceLocation::toString).map(StringValue::of));
```

File: src/main/java/carpet/script/api/WorldAccess.java
Comment (altrisi) [inline src/main/java/carpet/script/api/WorldAccess.java:1698]: Can this actually ever happen? If it shouldn't I'd remove it or throw a non-scarpet exception (like IllegalState/Argument)

**Review by altrisi — COMMENTED** [review]
Some comments. Are there many density functions usually in a world? Wondering if that's the case if it may be more useful to return a set? Though I don't know as I don't have a use case. Not blocking.

File: src/main/java/carpet/script/api/WorldAccess.java
Comment (Crec0) [reply src/main/java/carpet/script/api/WorldAccess.java:1698]: No. This should never happen. I did it just to make one less warning "densityFuction could be null"

**Comment by Crec0** [comment]
> Some comments. Are there many density functions usually in a world? Wondering if that's the case if it may be more useful to return a set? Though I don't know as I don't have a use case. Not blocking.

Yes. Many. You can boot the game and run command `script run compute_density_function(p~'pos')` to get the list of currently available density function. World generation datapacks add a lot of their own density functions too.

**Comment by altrisi** [comment]
A set I mean when returning the available noises (length 0), was wondering if checking "do available noises contain x" was a common thing.

File: src/main/java/carpet/script/api/WorldAccess.java
Comment (altrisi) [inline src/main/java/carpet/script/api/WorldAccess.java:1645]: Didn't realize: Removes the `minecraft` namespace, like other functions do, and removes a map call
```suggestion
            return ListValue.wrap(densityFunctionRegistry.keySet().stream().map(ValueConversions::of));
```

**Review by altrisi — APPROVED** [review]
Will merge after snapshot, just got to test it one last time after the merge

**Comment by Crec0** [comment]
> A set I mean when returning the available noises (length 0), was wondering if checking "do available noises contain x" was a common thing.

Oh i see. 0 names passed in is more of a debug thing instead of actual usage thing. Using a set or list is pretty much same effect. However I do see how set would be kinda better in case someone actually uses it in a loop with filters and stuff. I will make this change.

**Comment by altrisi** [comment]
If it's going to be debug only I'd keep it as a list (simpler to use), was wondering if an app may want to check that a data pack function is present for example.

Though maybe it should just check for the data pack to be present in that case...

File: src/main/java/carpet/script/api/WorldAccess.java
Comment (altrisi) [reply src/main/java/carpet/script/api/WorldAccess.java:1698]:
```suggestion
            if (densityFunction == null) // IDE thinks may be null as Registry.get is Nullable, but we checked it
                throw new IllegalStateException("density function can't be null here"); 
```

**Pull request merged**
